pkg/controller/restic_key_change_controller.go

```diff
@@ -0,0 +1,188 @@
+
+	namespace              string
+	secretLister           corev1listers.SecretLister
+	secretClient           corev1client.SecretsGetter
+	resticRepositoryClient arkv1client.ResticRepositoriesGetter
+	resticRepositoryLister listers.ResticRepositoryLister
+	repositoryManager      restic.RepositoryManager
+}
+
+// NewResticKeyChangeController creates a new restic key-change controller.
+func NewResticKeyChangeController(
+	logger logrus.FieldLogger,
+	namespace string,
+	secretInformer cache.SharedIndexInformer,
+	secretClient corev1client.SecretsGetter,
+	resticRepositoryInformer informers.ResticRepositoryInformer,
+	resticRepositoryClient arkv1client.ResticRepositoriesGetter,
+	repositoryManager restic.RepositoryManager,
+) Interface {
+	c := &resticKeyChangeController{
+		genericController:      newGenericController("restic-key-change", logger),
+		namespace:              namespace,
+		secretClient:           secretClient,
+		secretLister:           corev1listers.NewSecretLister(secretInformer.GetIndexer()),
+		resticRepositoryClient: resticRepositoryClient,
+		resticRepositoryLister: resticRepositoryInformer.Lister(),
+		repositoryManager:      repositoryManager,
+	}
+
+	c.syncHandler = c.processQueueItem
+	c.cacheSyncWaiters = append(c.cacheSyncWaiters, secretInformer.HasSynced, resticRepositoryInformer.Informer().HasSynced)
+
+	secretInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: c.enqueueIfKeyChange,
+			UpdateFunc: func(_, obj interface{}) {
+				c.enqueueIfKeyChange(obj)
+			},
+		},
+	)
+
+	return c
+}
+
+func shouldProcess(secret *corev1.Secret) bool {
+	// we need to process a secret if it's got data in either the "new-key"
+	// or "old-key" secret slots
+	if _, ok := secret.Data[restic.NewCredentialsKey]; ok {
+		return true
+	}
+
+	if _, ok := secret.Data[restic.OldCredentialsKey]; ok {
+		return true
+	}
+
+	return false
+}
+
+func (c *resticKeyChangeController) enqueueIfKeyChange(obj interface{}) {
+	secret := obj.(*corev1.Secret)
+
+	if shouldProcess(secret) {
+		c.enqueue(obj)
+	}
+}
+
+func (c *resticKeyChangeController) processQueueItem(key string) error {
+	c.logger.WithField("key", key).Debug("Running processQueueItem")
+
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		c.logger.WithField("key", key).WithError(errors.WithStack(err)).Error("error splitting queue key")
+		return nil
+	}
+	log := c.logger.WithField("namespace", ns).WithField("name", name)
+
+	secret, err := c.secretLister.Secrets(ns).Get(name)
+	if apierrors.IsNotFound(err) {
+		log.Debug("Unable to find secret")
+		return nil
+	}
+	if err != nil {
+		return errors.Wrapf(err, "error getting secret")
+	}
+
+	// Don't mutate the shared cache
+	secretCopy := secret.DeepCopy()
+
+	return c.processKeyChange(secretCopy, log)
+}
+
+func (c *resticKeyChangeController) processKeyChange(secret *corev1.Secret, log logrus.FieldLogger) error {
+	// check again if we should process because secret may have changed
```

Comment:
I would move this up to `processQueueItem` and do it before you make the deep copy